### ISSUE TITLE
Time complexity is higher than other implementations

### ISSUE THREAD

**Issue by Kipsora** [CyberZHG/keras-adaptive-softmax#2]
**Describe the Bug**

In [other implementations](https://github.com/kimiyoung/transformer-xl/blob/master/tf/model.py#L261-L264), by using labels to mask out those useless values we can further shrink the time complexity. But here in this repo the adaptive softmax has been implemented in a way that doesn't use labels. Suppose the vocabulary is divided into V = V1 + V2 + ... + Vk, model dimension is D, the scaling factor is s, and batch size is B which contains B1 words in V1, B2 words in V2, ..., and Bk words in Vk. Then current implementation in this repo is O(BD(V1 + k) + **B**DV2/s + **B**DV3/s^2 + ... + **B**DVk/s^(k-1)) but if the labels are present, we can further optimize the complexity to O(BD(V1 + k) + **B2**DV2/s + **B3**DV3/s^2 + ... + **Bk**DVk/s^(k-1)).

**Version Info**

* [x] I'm using the latest version

**Minimal Codes To Reproduce**

```python
import keras_adaptive_softmax

pass
```

**Comment by CyberZHG** [comment]
I think the Transformer-XL's implementation is also `B` times other factors as it uses `scatter_nd`?

**Comment by Kipsora** [comment]
@CyberZHG Yeah. But I think all `scatter_nd` together should only take `O(B*k)` (the size of [`target`](https://github.com/kimiyoung/transformer-xl/blob/master/tf/model.py#L258)) since it only involves with addition instead of multiplication.

**Comment by CyberZHG** [comment]
And I think the `gather` function does the masking work:

https://github.com/CyberZHG/keras-adaptive-softmax/blob/master/keras_adaptive_softmax/embedding.py#L182

**Comment by Kipsora** [comment]
> And I think the `gather` function does the masking work:
> 
> https://github.com/CyberZHG/keras-adaptive-softmax/blob/master/keras_adaptive_softmax/embedding.py#L182

For embedding it's correct, but for softmax it seems not. Sorry for the confusion.

**Comment by CyberZHG** [comment]
How do you know the labels before `softmax`?

**Comment by Kipsora** [comment]
There is no way to know the labels before `softmax` if we only care about the softmax values. But oftentimes the softmax layer is followed by a cross entropy layer, then the `softmax` can be computed with labels present. (So this issue should better be a feature request instead of a bug report)
